cmd/envtool/tests.go

```diff
@@ -17,24 +17,115 @@ package main
 import (
 	"bufio"
 	"bytes"
+	"encoding/json"
 	"fmt"
 	"io"
 	"os"
 	"os/exec"
 	"sort"
 	"strings"
+	"time"
 
 	"go.uber.org/zap"
 	"golang.org/x/exp/maps"
 
 	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
 )
 
+type testEvent struct {
+	Time           time.Time `json:"Time"`
+	Action         string    `json:"Action"`
+	Package        string    `json:"Package"`
+	Test           string    `json:"Test"`
+	Output         string    `json:"Output"`
+	ElapsedSeconds float64   `json:"Elapsed"`
+}
+
+func extractRootTestName(fullTestName string) string {
+	parts := strings.Split(fullTestName, "/")
+	return parts[0]
+}
+
+// execTestCommand runs test with given arguments.
+func execTestCommand(command string, args []string, totalTest int, logger *zap.SugaredLogger) error {
+	bin, err := exec.LookPath(command)
+	if err != nil {
+		return err
+	}
+	cmd := exec.Command(bin, args...)
+	logger.Debugf("Running %s", strings.Join(cmd.Args, " "))
+
+	cmd.Stderr = os.Stderr
+	p, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+
+	if err = cmd.Start(); err != nil {
+		return err
+	}
+
+	var testCounter int = 1
+	tested := make(map[string]bool)
+	d := json.NewDecoder(p)
+	d.DisallowUnknownFields()
+
+	for {
+		var event testEvent
+		if err = d.Decode(&event); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+
+		// skip package failures
+		if event.Test == "" {
+			continue
+		}
+
+		rootTestName := extractRootTestName(event.Test)
+		switch event.Action {
+		case "pass":
+			_, exists := tested[rootTestName]
+			if !exists {
```

Comment:
And what happens if it already exists? Can it happen?

Reply:
I have problem when only save it as string and test run `t.parallel()` back and forth testname overwrittern. Because of that I save it as map